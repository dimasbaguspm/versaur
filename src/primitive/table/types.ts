import type { HTMLAttributes, ReactNode, TableHTMLAttributes } from 'react'

/**
 * TableProps for the root Table component
 */
export interface TableProps extends TableHTMLAttributes<HTMLTableElement> {
  /** Table children (Header, Body, Footer, etc) */
  children: ReactNode
  /** Maximum number of columns in the table */
  columns: number
}

/**
 * Context value for Table
 */
export interface TableContextValue {
  columns: number
}

/**
 * TableHeaderProps for <Table.Header>
 */
export interface TableHeaderProps extends HTMLAttributes<HTMLDivElement> {
  children: ReactNode
}

/**
 * TableBodyProps for <Table.Body>
 */
export interface TableBodyProps extends HTMLAttributes<HTMLDivElement> {
  children: ReactNode
}

/**
 * TableFooterProps for <Table.Footer>
 */
export interface TableFooterProps extends HTMLAttributes<HTMLDivElement> {
  children: ReactNode
}

/**
 * TableRowProps for <Table.Row>
 */
export interface TableRowProps extends HTMLAttributes<HTMLDivElement> {
  children: ReactNode
}

/**
 * TableColumnProps for <Table.Column>
 * @property as - 'th' for header, 'td' for body/footer (default: 'td')
 * @property span - number of grid columns to span (1-12)
 */
export interface TableColumnProps extends HTMLAttributes<HTMLDivElement> {
  children: ReactNode
  as: 'th' | 'td'
  span: number // 1-12
  /** Horizontal alignment of cell content (left, center, right). Default: left */
  align?: 'left' | 'center' | 'right'
}

export type TableHeaderItemProps = Omit<TableColumnProps, 'as'>
export type TableRowItemProps = Omit<TableColumnProps, 'as'>
