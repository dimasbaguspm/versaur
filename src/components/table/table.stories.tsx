/**
 * Table stories for Versaur UI
 *
 * Demonstrates the Table compound component usage and variations
 */
import { Table } from './index'

export default {
  title: 'Components/Table',
  component: Table,
  parameters: {
    docs: {
      description: {
        component:
          'Compound Table component for accessible, styled, and composable table UI.',
      },
    },
  },
}

export const Basic = () => (
  <Table>
    <Table.Header>
      <Table.Column as='th' span={4}>
        Date
      </Table.Column>
      <Table.Column as='th' span={4}>
        Type
      </Table.Column>
      <Table.Column as='th' span={2}>
        Amount
      </Table.Column>
      <Table.Column as='th' span={2}>
        Action
      </Table.Column>
    </Table.Header>
    <Table.Body>
      <Table.Row>
        <Table.Column as='td' span={4}>
          2025-07-01
        </Table.Column>
        <Table.Column as='td' span={4}>
          Income
        </Table.Column>
        <Table.Column as='td' span={2}>
          $2,000
        </Table.Column>
        <Table.Column as='td' span={2}>
          <button>Edit</button>
        </Table.Column>
      </Table.Row>
      <Table.Row>
        <Table.Column as='td' span={4}>
          2025-07-02
        </Table.Column>
        <Table.Column as='td' span={4}>
          Expense
        </Table.Column>
        <Table.Column as='td' span={2}>
          -$500
        </Table.Column>
        <Table.Column as='td' span={2}>
          <button>Edit</button>
        </Table.Column>
      </Table.Row>
      <Table.Row>
        <Table.Column as='td' span={4}>
          2025-07-03
        </Table.Column>
        <Table.Column as='td' span={4}>
          Income
        </Table.Column>
        <Table.Column as='td' span={2}>
          $1,200
        </Table.Column>
        <Table.Column as='td' span={2}>
          <button>Edit</button>
        </Table.Column>
      </Table.Row>
    </Table.Body>
    <Table.Footer>
      <Table.Column as='td' span={8}>
        Total
      </Table.Column>
      <Table.Column as='td' span={4}>
        $2,700
      </Table.Column>
    </Table.Footer>
  </Table>
)

/**
 * ColumnSpan story demonstrates various column span (1-12) and mixed widths
 */
export const ColumnSpan = () => (
  <Table>
    <Table.Header>
      <Table.Column as='th' span={3}>
        Date
      </Table.Column>
      <Table.Column as='th' span={3}>
        Type
      </Table.Column>
      <Table.Column as='th' span={3}>
        Amount
      </Table.Column>
      <Table.Column as='th' span={3}>
        Action
      </Table.Column>
    </Table.Header>
    <Table.Body>
      <Table.Row>
        <Table.Column as='td' span={6}>
          2025-07-01
        </Table.Column>
        <Table.Column as='td' span={3}>
          Income
        </Table.Column>
        <Table.Column as='td' span={2}>
          $2,000
        </Table.Column>
        <Table.Column as='td' span={1}>
          <button>Edit</button>
        </Table.Column>
      </Table.Row>
      <Table.Row>
        <Table.Column as='td' span={3}>
          2025-07-02
        </Table.Column>
        <Table.Column as='td' span={6}>
          Expense
        </Table.Column>
        <Table.Column as='td' span={2}>
          -$500
        </Table.Column>
        <Table.Column as='td' span={1}>
          <button>Edit</button>
        </Table.Column>
      </Table.Row>
      <Table.Row>
        <Table.Column
          as='td'
          span={12}
          className='text-center font-semibold bg-[var(--color-tertiary-soft)]'
        >
          Spans all 12 columns
        </Table.Column>
      </Table.Row>
      <Table.Row>
        <Table.Column as='td' span={2}>
          A
        </Table.Column>
        <Table.Column as='td' span={2}>
          B
        </Table.Column>
        <Table.Column as='td' span={3}>
          C
        </Table.Column>
        <Table.Column as='td' span={5}>
          D
        </Table.Column>
      </Table.Row>
    </Table.Body>
  </Table>
)

/**
 * Alignment story demonstrates left, center, and right alignment in TableColumn
 */
export const Alignment = () => (
  <Table>
    <Table.Header>
      <Table.Column as='th' span={4} align='left'>
        Left
      </Table.Column>
      <Table.Column as='th' span={4} align='center'>
        Center
      </Table.Column>
      <Table.Column as='th' span={4} align='right'>
        Right
      </Table.Column>
    </Table.Header>
    <Table.Body>
      <Table.Row>
        <Table.Column as='td' span={4} align='left'>
          2025-07-01
        </Table.Column>
        <Table.Column as='td' span={4} align='center'>
          Centered
        </Table.Column>
        <Table.Column as='td' span={4} align='right'>
          $2,000
        </Table.Column>
      </Table.Row>
      <Table.Row>
        <Table.Column as='td' span={4}>
          Default (left)
        </Table.Column>
        <Table.Column as='td' span={4} align='center'>
          Centered
        </Table.Column>
        <Table.Column as='td' span={4} align='right'>
          Right
        </Table.Column>
      </Table.Row>
    </Table.Body>
  </Table>
)
