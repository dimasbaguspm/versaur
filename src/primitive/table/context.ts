import { createContext, useContext } from 'react'

import type { TableContextValue } from './types'

const TableContext = createContext<TableContextValue | null>(null)

export const TableProvider = TableContext.Provider

export function useTableContext() {
  const ctx = useContext(TableContext)
  if (!ctx) {
    throw new Error('Table compound components must be used within <Table>')
  }
  return ctx
}
