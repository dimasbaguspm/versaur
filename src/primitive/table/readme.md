# Table Component

Compound Table component for Versaur UI library. Provides accessible, styled, and composable table UI using the compound pattern.

## Usage


```tsx
<Table>
  <Table.Header>
    <Table.Column as="th" span={4} align="left">Date</Table.Column>
    <Table.Column as="th" span={4} align="center">Type</Table.Column>
    <Table.Column as="th" span={2} align="right">Amount</Table.Column>
    <Table.Column as="th" span={2}>Action</Table.Column>
  </Table.Header>
  <Table.Body>
    <Table.Row>
      <Table.Column as="td" span={4} align="left">2025-07-01</Table.Column>
      <Table.Column as="td" span={4} align="center">Income</Table.Column>
      <Table.Column as="td" span={2} align="right">$2,000</Table.Column>
      <Table.Column as="td" span={2}><button>Edit</button></Table.Column>
    </Table.Row>
    <Table.Row>
      <Table.Column as="td" span={4}>2025-07-02</Table.Column>
      <Table.Column as="td" span={4} align="center">Expense</Table.Column>
      <Table.Column as="td" span={2} align="right">-$500</Table.Column>
      <Table.Column as="td" span={2}><button>Edit</button></Table.Column>
    </Table.Row>
  </Table.Body>
  <Table.Footer>
    <Table.Column as="td" span={8} align="left">Total</Table.Column>
    <Table.Column as="td" span={2} align="right">$1,500</Table.Column>
    <Table.Column as="td" span={2}></Table.Column>
  </Table.Footer>
</Table>
```


- Use `as="th"` for header columns, `as="td"` (default) for body/footer columns.
- Use the `span` prop to control column width (1-12 grid system).
- Use the `align` prop (`left` | `center` | `right`, default: `left`) to control horizontal alignment of cell content.
- Sorting and other stateful features are managed via context.
