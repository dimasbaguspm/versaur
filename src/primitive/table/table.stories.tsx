/**
 * Table stories for Versaur UI
 *
 * Demonstrates the Table compound component usage and variations
 */
import { Pen, Trash } from 'lucide-react'
import { ButtonIcon } from '../button-icon'
import { Table } from './index'

export default {
  title: 'Primitive/Table',
  component: Table,
  parameters: {
    docs: {
      description: {
        component:
          'Compound Table component for accessible, styled, and composable table UI.',
      },
    },
  },
}

export const Basic = () => (
  <Table columns={12}>
    <Table.Header>
      <Table.HeaderItem span={4}>Date</Table.HeaderItem>
      <Table.HeaderItem span={4}>Type</Table.HeaderItem>
      <Table.HeaderItem span={2}>Amount</Table.HeaderItem>
      <Table.HeaderItem span={2}>Action</Table.HeaderItem>
    </Table.Header>
    <Table.Body>
      <Table.Row>
        <Table.RowItem span={4}>2025-07-01</Table.RowItem>
        <Table.RowItem span={4}>Income</Table.RowItem>
        <Table.RowItem span={2}>$2,000</Table.RowItem>
        <Table.RowItem span={2}>
          <ButtonIcon as={Pen} variant='ghost' size='sm' aria-label='Edit' />
          <ButtonIcon
            as={Trash}
            variant='danger-ghost'
            size='sm'
            aria-label='Delete'
          />
        </Table.RowItem>
      </Table.Row>
      <Table.Row>
        <Table.RowItem span={4}>2025-07-02</Table.RowItem>
        <Table.RowItem span={4}>Expense</Table.RowItem>
        <Table.RowItem span={2}>-$500</Table.RowItem>
        <Table.RowItem span={2}>
          <ButtonIcon as={Pen} variant='ghost' size='sm' aria-label='Edit' />
          <ButtonIcon
            as={Trash}
            variant='danger-ghost'
            size='sm'
            aria-label='Delete'
          />
        </Table.RowItem>
      </Table.Row>
      <Table.Row>
        <Table.RowItem span={4}>2025-07-03</Table.RowItem>
        <Table.RowItem span={4}>Income</Table.RowItem>
        <Table.RowItem span={2}>$1,200</Table.RowItem>
        <Table.RowItem span={2}>
          <ButtonIcon as={Pen} variant='ghost' size='sm' aria-label='Edit' />
          <ButtonIcon
            as={Trash}
            variant='danger-ghost'
            size='sm'
            aria-label='Delete'
          />
        </Table.RowItem>
      </Table.Row>
    </Table.Body>
    <Table.Footer>
      <Table.RowItem span={8}>Total</Table.RowItem>
      <Table.RowItem span={4}>$2,700</Table.RowItem>
    </Table.Footer>
  </Table>
)

/**
 * ColumnSpan story demonstrates various column span (1-12) and mixed widths
 */
export const ColumnSpan = () => (
  <Table columns={12}>
    <Table.Header>
      <Table.HeaderItem span={3}>Date</Table.HeaderItem>
      <Table.HeaderItem span={3}>Type</Table.HeaderItem>
      <Table.HeaderItem span={3}>Amount</Table.HeaderItem>
      <Table.HeaderItem span={3}>Action</Table.HeaderItem>
    </Table.Header>
    <Table.Body>
      <Table.Row>
        <Table.RowItem span={6}>2025-07-01</Table.RowItem>
        <Table.RowItem span={3}>Income</Table.RowItem>
        <Table.RowItem span={2}>$2,000</Table.RowItem>
        <Table.RowItem span={1}>
          <ButtonIcon as={Pen} variant='ghost' size='sm' aria-label='Edit' />
        </Table.RowItem>
      </Table.Row>
      <Table.Row>
        <Table.RowItem span={3}>2025-07-02</Table.RowItem>
        <Table.RowItem span={6}>Expense</Table.RowItem>
        <Table.RowItem span={2}>-$500</Table.RowItem>
        <Table.RowItem span={1}>
          <ButtonIcon as={Pen} variant='ghost' size='sm' aria-label='Edit' />
        </Table.RowItem>
      </Table.Row>
      <Table.Row>
        <Table.RowItem
          span={12}
          className='text-center font-semibold bg-[var(--color-tertiary-soft)]'
        >
          Spans all 12 columns
        </Table.RowItem>
      </Table.Row>
      <Table.Row>
        <Table.RowItem span={2}>A</Table.RowItem>
        <Table.RowItem span={2}>B</Table.RowItem>
        <Table.RowItem span={3}>C</Table.RowItem>
        <Table.RowItem span={5}>D</Table.RowItem>
      </Table.Row>
    </Table.Body>
  </Table>
)

/**
 * Alignment story demonstrates left, center, and right alignment in TableColumn
 */
export const Alignment = () => (
  <Table columns={12}>
    <Table.Header>
      <Table.HeaderItem span={4} align='left'>
        Left
      </Table.HeaderItem>
      <Table.HeaderItem span={4} align='center'>
        Center
      </Table.HeaderItem>
      <Table.HeaderItem span={4} align='right'>
        Right
      </Table.HeaderItem>
    </Table.Header>
    <Table.Body>
      <Table.Row>
        <Table.RowItem span={4} align='left'>
          2025-07-01
        </Table.RowItem>
        <Table.RowItem span={4} align='center'>
          Centered
        </Table.RowItem>
        <Table.RowItem span={4} align='right'>
          $2,000
        </Table.RowItem>
      </Table.Row>
      <Table.Row>
        <Table.RowItem span={4}>Default (left)</Table.RowItem>
        <Table.RowItem span={4} align='center'>
          Centered
        </Table.RowItem>
        <Table.RowItem span={4} align='right'>
          Right
        </Table.RowItem>
      </Table.Row>
    </Table.Body>
  </Table>
)
