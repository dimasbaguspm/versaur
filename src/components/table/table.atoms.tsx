import { forwardRef } from 'react'
import type {
  TableHeaderProps,
  TableBodyProps,
  TableFooterProps,
  TableRowProps,
  TableColumnProps,
} from './types'
import { cn } from '@/utils/cn'
import { useTableContext } from './context'
import { getRowSpanClass, getTableColumnClass } from './helpers'

/**
 * Table.Header renders the table header row (grid-based, ARIA role)
 */
const TableHeader = forwardRef<HTMLDivElement, TableHeaderProps>(
  ({ children, className, ...props }, ref) => {
    const { maxColumns } = useTableContext()
    return (
      <div
        role='rowgroup'
        className={cn('bg-neutral-soft', className)}
        {...props}
        ref={ref}
      >
        <div
          role='row'
          className={cn('grid gap-4', getTableColumnClass(maxColumns))}
        >
          {children}
        </div>
      </div>
    )
  }
)

/**
 * Table.Body renders the table body (grid-based, ARIA role)
 */
const TableBody = forwardRef<HTMLDivElement, TableBodyProps>(
  ({ children, className, ...props }, ref) => (
    <div role='rowgroup' ref={ref} className={cn(className)} {...props}>
      {children}
    </div>
  )
)

/**
 * Table.Footer renders the table footer row (grid-based, ARIA role)
 */
const TableFooter = forwardRef<HTMLDivElement, TableFooterProps>(
  ({ children, className, ...props }, ref) => {
    const { maxColumns } = useTableContext()
    return (
      <div
        role='rowgroup'
        className={cn('bg-neutral-soft', className)}
        {...props}
        ref={ref}
      >
        <div
          role='row'
          className={cn('grid gap-4', getTableColumnClass(maxColumns))}
        >
          {children}
        </div>
      </div>
    )
  }
)

/**
 * Table.Row renders a table row (grid-based, ARIA role)
 */
const TableRow = forwardRef<HTMLDivElement, TableRowProps>(
  ({ children, className, ...props }, ref) => {
    const { maxColumns } = useTableContext()

    return (
      <div
        role='row'
        className={cn(
          'grid gap-4',
          getTableColumnClass(maxColumns),
          'border-b border-border last:border-0',
          className
        )}
        ref={ref}
        {...props}
      >
        {children}
      </div>
    )
  }
)

/**
 * Table.Column renders a table cell (grid-based, ARIA role)
 */
const TableColumn = forwardRef<HTMLDivElement, TableColumnProps>(
  ({ as = 'td', span, align = 'left', children, className, ...props }, ref) => {
    const role = as === 'th' ? 'columnheader' : 'cell'
    let alignClass = 'text-left'
    if (align === 'center') alignClass = 'text-center'
    else if (align === 'right') alignClass = 'text-right'

    return (
      <div
        ref={ref}
        role={role}
        className={cn(
          'px-4 py-2',
          '[&:not(:last-child)]:border-r [&:not(:last-child)]:border-border',
          getRowSpanClass(span),
          alignClass,
          className
        )}
        {...props}
      >
        {children}
      </div>
    )
  }
)

export { TableHeader, TableBody, TableFooter, TableRow, TableColumn }
