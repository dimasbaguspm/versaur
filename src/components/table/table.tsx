import { useMemo, forwardRef } from 'react'
import type { TableContextValue, TableProps } from './types'
import { cn } from '@/utils/cn'
import { TableProvider } from './context'
import {
  TableBody,
  TableColumn,
  TableFooter,
  TableHeader,
  TableRow,
} from './table.atoms'

const TableRoot = forwardRef<HTMLDivElement, TableProps>(
  ({ children, className, maxColumns = 12, ...props }, ref) => {
    const value = useMemo(
      () => ({ maxColumns }),
      [maxColumns]
    ) satisfies TableContextValue

    return (
      <TableProvider value={value}>
        <div
          ref={ref}
          role='table'
          className={cn(
            'w-full overflow-x-auto rounded-lg border border-border',
            className
          )}
          {...props}
        >
          {children}
        </div>
      </TableProvider>
    )
  }
)

export const Table = Object.assign(TableRoot, {
  Header: TableHeader,
  Body: TableBody,
  Footer: TableFooter,
  Row: TableRow,
  Column: TableColumn,
})
