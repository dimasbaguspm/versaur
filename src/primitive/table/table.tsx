import { forwardRef } from 'react'
import type { TableProps } from './types'
import { cn } from '@/utils/cn'
import { TableProvider } from './context'
import {
  TableBody,
  TableFooter,
  TableHeader,
  TableHeaderItem,
  TableRow,
  TableRowItem,
} from './table.atoms'

const TableRoot = forwardRef<HTMLDivElement, TableProps>(
  ({ children, className, columns = 12, ...props }, ref) => (
    <TableProvider value={{ columns }}>
      <div
        ref={ref}
        role='table'
        className={cn(
          // Versaur design system: border, background, shadow, rounded, spacing
          'w-full overflow-x-auto border border-border bg-background rounded-lg ',
          'text-foreground',
          'sm:rounded-xl',
          'transition-colors',
          className
        )}
        {...props}
      >
        {children}
      </div>
    </TableProvider>
  )
)

export const Table = Object.assign(TableRoot, {
  Header: TableHeader,
  HeaderItem: TableHeaderItem,
  Body: TableBody,
  Footer: TableFooter,
  Row: TableRow,
  RowItem: TableRowItem,
})
