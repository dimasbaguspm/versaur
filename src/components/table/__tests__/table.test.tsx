import { composeStories } from '@storybook/react'
import * as stories from '../table.stories'
import { render, screen } from '@testing-library/react'

describe('Table', () => {
  const { Basic, Alignment } = composeStories(stories)

  it('renders table structure and content correctly', () => {
    const { asFragment } = render(<Basic />)
    expect(asFragment()).toMatchSnapshot()
    expect(screen.getByRole('table')).toBeInTheDocument()
    expect(screen.getByText('Date')).toBeInTheDocument()
    // There are two 'Income' cells (header and body), check at least one exists
    expect(screen.getAllByText('Income').length).toBeGreaterThan(0)
    expect(screen.getByText('Total')).toBeInTheDocument()
    expect(screen.getAllByRole('row')).toHaveLength(5) // header, 3 body, footer
    expect(screen.getAllByRole('cell').length).toBeGreaterThan(0)
  })

  it('renders alignment correctly for TableColumn', () => {
    render(<Alignment />)
    // Check left, center, right header
    // Header alignment
    const [leftHeader, centerHeader, rightHeader] = [
      ...screen.getAllByText('Left'),
      ...screen.getAllByText('Center'),
      ...screen.getAllByText('Right'),
    ]
    expect(leftHeader.className).toMatch(/text-left/)
    expect(centerHeader.className).toMatch(/text-center/)
    expect(rightHeader.className).toMatch(/text-right/)
    // Check left, center, right body
    // Body alignment
    const [dateCell] = screen.getAllByText('2025-07-01')
    const [centeredCell] = screen.getAllByText('Centered')
    const [rightCell] = screen.getAllByText('$2,000')
    expect(dateCell.className).toMatch(/text-left/)
    expect(centeredCell.className).toMatch(/text-center/)
    expect(rightCell.className).toMatch(/text-right/)
    // Default align (should be left)
    const [defaultLeftCell] = screen.getAllByText('Default (left)')
    expect(defaultLeftCell.className).toMatch(/text-left/)
  })
})
