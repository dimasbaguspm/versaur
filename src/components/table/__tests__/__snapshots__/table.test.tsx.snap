// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table > renders table structure and content correctly 1`] = `
<DocumentFragment>
  <div
    class="w-full overflow-x-auto rounded-lg border border-border"
    role="table"
  >
    <div
      class="bg-neutral-soft"
      role="rowgroup"
    >
      <div
        class="grid gap-4 grid-cols-12"
        role="row"
      >
        <div
          class="px-4 py-2 [&:not(:last-child)]:border-r [&:not(:last-child)]:border-border col-span-4 text-left"
          role="columnheader"
        >
          Date
        </div>
        <div
          class="px-4 py-2 [&:not(:last-child)]:border-r [&:not(:last-child)]:border-border col-span-4 text-left"
          role="columnheader"
        >
          Type
        </div>
        <div
          class="px-4 py-2 [&:not(:last-child)]:border-r [&:not(:last-child)]:border-border col-span-2 text-left"
          role="columnheader"
        >
          Amount
        </div>
        <div
          class="px-4 py-2 [&:not(:last-child)]:border-r [&:not(:last-child)]:border-border col-span-2 text-left"
          role="columnheader"
        >
          Action
        </div>
      </div>
    </div>
    <div
      class=""
      role="rowgroup"
    >
      <div
        class="grid gap-4 grid-cols-12 border-b border-border last:border-0"
        role="row"
      >
        <div
          class="px-4 py-2 [&:not(:last-child)]:border-r [&:not(:last-child)]:border-border col-span-4 text-left"
          role="cell"
        >
          2025-07-01
        </div>
        <div
          class="px-4 py-2 [&:not(:last-child)]:border-r [&:not(:last-child)]:border-border col-span-4 text-left"
          role="cell"
        >
          Income
        </div>
        <div
          class="px-4 py-2 [&:not(:last-child)]:border-r [&:not(:last-child)]:border-border col-span-2 text-left"
          role="cell"
        >
          $2,000
        </div>
        <div
          class="px-4 py-2 [&:not(:last-child)]:border-r [&:not(:last-child)]:border-border col-span-2 text-left"
          role="cell"
        >
          <button>
            Edit
          </button>
        </div>
      </div>
      <div
        class="grid gap-4 grid-cols-12 border-b border-border last:border-0"
        role="row"
      >
        <div
          class="px-4 py-2 [&:not(:last-child)]:border-r [&:not(:last-child)]:border-border col-span-4 text-left"
          role="cell"
        >
          2025-07-02
        </div>
        <div
          class="px-4 py-2 [&:not(:last-child)]:border-r [&:not(:last-child)]:border-border col-span-4 text-left"
          role="cell"
        >
          Expense
        </div>
        <div
          class="px-4 py-2 [&:not(:last-child)]:border-r [&:not(:last-child)]:border-border col-span-2 text-left"
          role="cell"
        >
          -$500
        </div>
        <div
          class="px-4 py-2 [&:not(:last-child)]:border-r [&:not(:last-child)]:border-border col-span-2 text-left"
          role="cell"
        >
          <button>
            Edit
          </button>
        </div>
      </div>
      <div
        class="grid gap-4 grid-cols-12 border-b border-border last:border-0"
        role="row"
      >
        <div
          class="px-4 py-2 [&:not(:last-child)]:border-r [&:not(:last-child)]:border-border col-span-4 text-left"
          role="cell"
        >
          2025-07-03
        </div>
        <div
          class="px-4 py-2 [&:not(:last-child)]:border-r [&:not(:last-child)]:border-border col-span-4 text-left"
          role="cell"
        >
          Income
        </div>
        <div
          class="px-4 py-2 [&:not(:last-child)]:border-r [&:not(:last-child)]:border-border col-span-2 text-left"
          role="cell"
        >
          $1,200
        </div>
        <div
          class="px-4 py-2 [&:not(:last-child)]:border-r [&:not(:last-child)]:border-border col-span-2 text-left"
          role="cell"
        >
          <button>
            Edit
          </button>
        </div>
      </div>
    </div>
    <div
      class="bg-neutral-soft"
      role="rowgroup"
    >
      <div
        class="grid gap-4 grid-cols-12"
        role="row"
      >
        <div
          class="px-4 py-2 [&:not(:last-child)]:border-r [&:not(:last-child)]:border-border col-span-8 text-left"
          role="cell"
        >
          Total
        </div>
        <div
          class="px-4 py-2 [&:not(:last-child)]:border-r [&:not(:last-child)]:border-border col-span-4 text-left"
          role="cell"
        >
          $2,700
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
